AutoSitemap_body.php

```diff
@@ -89,34 +89,88 @@
 
 if (!isset($wgAutoSitemap["footer"]           )) $wgAutoSitemap["footer"]              = "\n</urlset>";
 
+if (!isset($wgAutoSitemap["min_age"]          )) $wgAutoSitemap["min_age"]             = 3600; // One hour in seconds
+
 class AutoSitemap {
     static public function writeSitemap() {
         global $wgAutoSitemap;
 
+        $filename = $wgAutoSitemap["filename"];
+        $min_age = $wgAutoSitemap["min_age"];
+
+        $mtime = filemtime($filename);
+        if ($mtime !== FALSE && time() - $mtime < $min_age) {
+           // Sitemap is young, no need to update.
+           return;
+        }
+
         $server       = $wgAutoSitemap["server"];
-        $filename     = $wgAutoSitemap["filename"];
         $tmp_filename = $filename.'.tmp'.bin2hex(random_bytes(16)).'.tmp';
 
-        $file_handle = fopen($tmp_filename, 'w') or die('Cannot write to '.$tmp_filename.'.');
+        $file_handle = fopen($tmp_filename, 'w');
+        if ($file_handle === FALSE) {
+           error_log("Couldn't fopen file: $tmp_filename");
+           return;
+        }
 
         $dbr = wfGetDB(DB_REPLICA);
         $res = $dbr->query(self::getSQL());
 
-        $count = $res->numRows();
-        $pos   = 0;
-
-        $data = $wgAutoSitemap["header"];
-        while($row = $res->fetchObject()) {
-            $data .= self::formatResult($server, $row, $pos, $count);
-            ++$pos;
+        // Sitemaps are limited to 50,000 URLs and 50 MB.
+        $entries    = 0;
+        $bytes      = 0;
+        $maxEntries = 50000;
+        $maxBytes   = 50000000;
+
+        /* Note:
+         *
+         * Actually, https://www.sitemaps.org/protocol.html says the limit is 52,428,800 bytes,
+         * which is 50 MiB (MiB = 1024 * 1024), but it uses the old notation of "MB" which has
+         * since been standardized in ISO to mean 1000 * 1000.  We use the lower limit just in
+         * case, because some systems may interpret the "50MB" to mean 50 * 1000 * 1000.
+         *
+         * It's very unlikely to reach the size limit before reaching the 50k URL limit anyway,
+         * unless there are tons of *extremely* long URLs in a sitemap.
+         */
+
+        $error = FALSE;
+
+        try {
+            $bytes += self::write($file_handle, $wgAutoSitemap["header"]);
+            while($row = $res->fetchObject()) {
+                if ($entries == $maxEntries) {
+                    throw new Exception("Sitemap URL limit exceeded, aborting.");
+                }
+                $entries += 1;
+                $bytes += self::write($file_handle, self::formatResult($server, $row));
+            }
+            $bytes += self::write($file_handle, $wgAutoSitemap["footer"]);
+            if ($bytes > $maxBytes) {
+                throw new Exception("Sitemap size limit exceeded, aborting.");
+            }
+        } catch (Exception $e) {
+            error_log("Writing sitemap failed: $e");
+            $error = TRUE;
```

Comment:
So if number of items exceeds the limit, sitemap will not be generated at all?

Reply:
Correct. It's probably better to stick with a valid sitemap that respects the limits instead of overriding it with an invalid one.

Reply:
Pages count usually increase, not decrease, so after reaching this limit sitemap will stop generating at all

Reply:
What alternative would you propose?

Reply:
Just show warning/error while generating sitemap file. Site admin should take care of this

Reply:
Is there a better way to do that than to use `error_log()`?

Reply:
Oh, you mean log error but generate it anyway? I think that would be bad, because now your sitemap will be invalid...

Reply:
If extension which generates a sitemap just stops updating it and shows an error, then it leads to shorter reaction time from site admin, but worse user experience.

It's better to show error but keep updating the file, until admin will move to other way if generating the sitemap. Adding an URL to documentation which explains the reason of this error and proposes some solution is even better.

To avoid generating a sitemap file which cannot be parsed by search engines, we can lower the limit which triggers this error message to something about 80-90% of the standard limit. Maybe even add several levels, for example warning message for 80% and error for 90%.